pkg/resource/reference_resolver.go

```diff
@@ -0,0 +1,222 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resource
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	apierrs "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/crossplaneio/crossplane-runtime/apis/core/v1alpha1"
+)
+
+const (
+	attributeReferencerTagName = "attribute_referencer"
+)
+
+// resourceNotReadyErr is a custom error used for references validation
+type resourceNotReadyErr struct{ msg string }
+
+func newResourceNotReadyErr(format string, args ...interface{}) *resourceNotReadyErr {
+	return &resourceNotReadyErr{fmt.Sprintf(format, args...)}
+}
+
+func (r *resourceNotReadyErr) Error() string {
+	return r.msg
+}
+
+// LocalObjectGetter is a function that is used to retrieve local object references
+type LocalObjectGetter func(ref corev1.LocalObjectReference, obj runtime.Object) error
+
+// AttributeReferencer is an interface for referencing and resolving
+// cross-resource attribute references. See
+// https://github.com/crossplaneio/crossplane/blob/master/design/one-pager-cross-resource-referencing.md
+// for more information
+type AttributeReferencer interface {
+	// Build retrieves and builds the referenced attribute
+	Build(getter LocalObjectGetter) (value string, err error)
+
+	// Assign accepts a resource object, and assigns the given value to the corresponding property
+	Assign(mgd Managed, value string) error
+}
+
+// A ManagedReferenceResolver resolves the references to other managed resources
+type ManagedReferenceResolver interface {
+	ResolveReferences(context.Context, Managed) error
+}
+
+// ReferenceResolver resolves implements ManagedReferenceResolver interface
+type ReferenceResolver struct {
+	reader client.Reader
+}
+
+// NewReferenceResolver returns a new ReferenceResolver
+func NewReferenceResolver(c client.Reader) *ReferenceResolver {
+	return &ReferenceResolver{c}
+}
+
+// ResolveReferences resolves references made to other managed resources
+func (r *ReferenceResolver) ResolveReferences(ctx context.Context, mgd Managed) (err error) {
+	// this recovers from potential panics during execution of
+	// AttributeReferencer methods
+	defer func() {
+		if r := recover(); r != nil {
+			err = errors.Errorf("paniced while resolving references: %v", r)
+		}
+	}()
+
+	referencers, err := findRefResolverFields(mgd, false)
+	if err != nil {
+		return err
+	}
+
+	getter := r.getterFunc(ctx, mgd)
+	err = validateReferencesAreReady(getter, referencers)
+	if err != nil {
+		return err
+	}
+
+	for _, referencer := range referencers {
+		val, err := referencer.Build(getter)
+		if err != nil {
+			return err
+		}
+
+		err = referencer.Assign(mgd, val)
+		if err != nil {
+			return err
+		}
```

Comment:
https://godoc.org/github.com/pkg/errors

While I don't think we necessarily need to wrap _every_ error, I think it could be useful to wrap the errors here, and possibly in a few other conditions introduced in this PR, to provide a little more context. e.g. Here I think an error string like "could not assign reference: reason" vs "could not build reference: reason" would be illustrative.

https://dave.cheney.net/2016/04/27/dont-just-check-errors-handle-them-gracefully explains the rationale in more detail  under "annotating errors".

Reply:
I very much agree with this, and I'll add context. I have worked with the `pkg/errors` you linked, and one issue is that to `Errorf` resets the error stack where it is called. This is fine for 
 capturing external errors for the first time, but using it later will rewrite the stack. The reason I didn't use this was because I didn't know where we stand regarding this.

It's worth exploring some options and conventions here. Go 1.13 also has added the notion of native [error wrappting](https://golang.org/doc/go1.13#error_wrapping)

Reply:
> Go 1.13 also has added the notion of native error wrappting

I think I personally prefer the `pkg/errors` API, but admittedly haven't spent enough time understanding the new support. My preference would be to stick with the established convention of using `pkg/errors` for the time being.